File: problems/0047-permutations-ii/solution.py
from typing import List

class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        """
        0047. Permutations II

        Given a list of integers `nums` (which may contain duplicates), return all unique permutations.

        Approach:
        - Sort the input list to ensure that duplicates are adjacent.
        - Use backtracking to generate permutations.
        - Track used indices to avoid repeating elements.
        - Skip duplicates by ensuring that for nums[i] == nums[i - 1], we only use it
          if the previous identical element has been used in the current branch.

        Time complexity: O(N*N!),
        where N is the number of elements in nums. There are at most N! permutations.
        Pruning duplicate branches reduces this significantly in practice.

        Space complexity: O(N),
        for recursion and tracking used elements.
        """

        nums.sort()  # Sort to make duplicate detection easier.
        res = []
        used = [False] * len(nums)

        def backtrack(path: List[int]) -> None:
            if len(path) == len(nums):
                res.append(path[:])
                return

            # Try each unused number at the current level.
            for i in range(len(nums)):
                if used[i]:
                    continue

                # Skip duplicates: if current is same as previous and previous was not used.
                if i > 0 and nums[i] == nums[i - 1] and not used[i - 1]:
                    continue

                # Choose nums[i].
                used[i] = True
                path.append(nums[i])

                # Recurse with nums[i] included.
                backtrack(path)

                # Backtrack.
                path.pop()
                used[i] = False

        backtrack([])
        return res

File: problems/0047-permutations-ii/solution.cpp
#include <vector>
#include <algorithm>
#include <functional>

using namespace std;

class Solution {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        /*
        0047. Permutations II

        Given a list of integers nums (which may contain duplicates), return all unique permutations.

        Approach:
        - Sort the input to ensure duplicates are adjacent.
        - Use backtracking to explore all permutations.
        - Track which elements have been used with a boolean array.
        - Skip over duplicate elements by ensuring that if nums[i] == nums[i - 1],
          we only use nums[i] if nums[i - 1] has been used in the current path.

        Time complexity: O(N*N!), where N is the number of elements in nums.
        There are at most N! permutations, but pruning avoids redundant ones.

        Space complexity: O(N), for the recursion stack and the used array.
        */

        sort(nums.begin(), nums.end());  // Sort to detect duplicates.
        vector<vector<int>> res;
        vector<int> path;
        vector<bool> used(nums.size(), false);

        function<void()> backtrack = [&]() {
            // If the path is complete, add a copy to results.
            if (path.size() == nums.size()) {
                res.push_back(path);
                return;
            }

            // Explore choices for the current position.
            for (int i = 0; i < nums.size(); i++) {
                if (used[i]) continue;  // Skip if already used.

                // Skip duplicates: only use the first unused duplicate at each level.
                if (i > 0 && nums[i] == nums[i - 1] && !used[i - 1]) continue;

                // Choose nums[i].
                used[i] = true;
                path.push_back(nums[i]);

                // Recurse.
                backtrack();

                // Backtrack.
                path.pop_back();
                used[i] = false;
            }
        };

        backtrack();  // Start the recursion.
        return res;
    }
};

